qiskit/circuit/quantumcircuit.py

```diff
@@ -1406,6 +1529,273 @@ def _update_parameter_table(self, instruction: CircuitInstruction):
+
+        Args:
+            name: the name of the variable to retrieve.
+            default: if given, this value will be returned if the variable is not present.  If it
+                is not given, a :exc:`KeyError` is raised instead.
+
+        Returns:
+            The corresponding variable.
+
+        Raises:
+            KeyError: if no default is given, but the variable does not exist.
+
+        Examples:
+            Retrieve a variable by name from a circuit::
+
+                from qiskit.circuit import QuantumCircuit
+
+                # Create a circuit and create a variable in it.
+                qc = QuantumCircuit()
+                my_var = qc.add_var("my_var", False)
+
+                # We can use 'my_var' as a variable, but let's say we've lost the Python object and
+                # need to retrieve it.
+                my_var_again = qc.get_var("my_var")
+
+                assert my_var is my_var_again
+
+            Get a variable from a circuit by name, returning some default if it is not present::
+
+                assert qc.get_var("my_var", None) is my_var
+                assert qc.get_var("unknown_variable", None) is None
+        """
+
+        if (out := self._vars_local.get(name)) is not None:
+            return out
+        if (out := self._vars_capture.get(name)) is not None:
+            return out
+        if (out := self._vars_input.get(name)) is not None:
+            return out
+        if default is Ellipsis:
+            raise KeyError(f"no variable named '{name}' is present")
+        return default
+
+    def has_var(self, name_or_var: str | expr.Var, /) -> bool:
+        """Check whether a variable is defined in this scope.
+
+        Args:
+            name_or_var: the variable, or name of a variable to check.  If this is a
+                :class:`.expr.Var` node, the variable must be exactly the given one for this
+                function to return ``True``.
+
+        Returns:
+            whether a matching variable is present.
+
+        See also:
+            :meth:`QuantumCircuit.get_var`: retrieve a named variable from a circuit.
+        """
+        if isinstance(name_or_var, str):
+            return self.get_var(name_or_var, None) is not None
+        return self.get_var(name_or_var.name, None) == name_or_var
+
+    def _prepare_new_var(
+        self, name_or_var: str | expr.Var, type_: types.Type | None, /
+    ) -> expr.Var:
+        """The common logic for preparing and validating a new :class:`~.expr.Var` for the circuit.
+
+        The given ``type_`` can be ``None`` if the variable specifier is already a :class:`.Var`,
+        and must be a :class:`~.types.Type` if it is a string.  The argument is ignored if the given
+        first argument is a :class:`.Var` already.
+
+        Returns the validated variable, which is guaranteed to be safe to add to the circuit."""
+        if isinstance(name_or_var, str):
+            var = expr.Var.new(name_or_var, type_)
```

Comment:
IIRC #10944 didn't add any validation of `type`. Is there something to guard against `name_or_var` being a string and `type_` being `None`, or is the assumption that the callers will always validate this? I was thinking we might want to just do:

```suggestion
            if type_ is None:
                raise CircuitError("Can't add a Var by string without a type")
            var = expr.Var.new(name_or_var, type_)
```

because you could call `add_uninitialized_var` with a string ignoring the type hint and I'm not sure it would error.

Reply:
Done in 39d7e3219c.  I've generally avoided type-checking everything because I don't want to cause slowdown by 1000 paper cuts (things like `copy` would have to suffer through unnecessary checking a whole lot of times), and because in Python it's always possible to make things do dodgy things unintentionally if you play havoc with the typing, but this is an easy change that would catch an understand user error with very very little overhead.